Allow rolebinding/clusterrolebinding with explicit bind permission check

liggitt opened the pull request:
Previously, a user could only create/update a role binding if they already had all the permissions contained in the referenced role (at the same scope as the role binding) via an existing RBAC role grant.

This PR also allows them to create/update a role binding if they’ve been given explicit permission to perform the `bind` verb on the referenced role. When using multiple authorizers (`--authorization-mode=RBAC,WebHook`, etc), that `bind` permission can be granted by any of the configured authorizers. This lets a non-RBAC authorizer indicate a user is allowed to bind specific RBAC roles.

For example, if “user-1” does not have the ability to list secrets in a namespace, they cannot create a RoleBinding to a role that grants that permission (like the `admin`, or `edit` role). This cluster role and role binding would now allow “user-1” to grant other users the `admin` and `edit` roles in the “user-1-namespace” namespace:

```yaml
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: role-grantor
rules:
- apiGroups: ["rbac.authorization.k8s.io"]
  resources: ["rolebindings"]
  verbs: ["create"]
- apiGroups: ["rbac.authorization.k8s.io"]
  resources: ["clusterroles"]
  verbs: ["bind"]
  resourceNames: ["admin","edit","view"]
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: RoleBinding
metadata:
  name: role-grantor-binding
  namespace: user-1-namespace
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: role-grantor
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: User
  name: user-1
```

Fixes https://github.com/kubernetes/kubernetes/issues/39176
Fixes https://github.com/kubernetes/kubernetes/issues/39258

Allows creating/updating a rolebinding/clusterrolebinding if the user has explicitly been granted permission to perform the "bind" verb against the referenced role/clusterrole (previously, they could only bind if they already had all the permissions in the referenced role via an RBAC role themselves)

```release-note
To create or update an RBAC RoleBinding or ClusterRoleBinding object, a user must:
1. Be authorized to make the create or update API request
2. Be allowed to bind the referenced role, either by already having all of the permissions contained in the referenced role, or by having the "bind" permission on the referenced role.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39383)
<!-- Reviewable:end -->

liggitt commented:
cc @cjcullen @deads2k @kubernetes/sig-auth-misc

liggitt committed:
Remove kubekins as cluster-admin

cjcullen commented on pkg/registry/rbac/clusterrolebinding/policybased/storage.go:
```diff
@@ -63,6 +66,12 @@ func (s *Storage) Update(ctx genericapirequest.Context, name string, obj rest.Up
 		clusterRoleBinding := obj.(*rbac.ClusterRoleBinding)
 
+		// if we're explicitly authorized to bind this clusterrole, return
+		if canBind, _, _ := rbacregistry.AuthorizeBinding(ctx, clusterRoleBinding.RoleRef, clusterRoleBinding.Namespace, s.authorizer); canBind {
```
Is it worth doing something else (at least logging) if we failed to get an explicit "no" here?

liggitt replied:
not sure, I don't think the explicit grant will be as common as just inheriting permissions, so I'd expect this check to fail most of the time, but it's cheaper than the covers check, so I want to do it first.

cjcullen replied:
I'm fine w/ not logging the `canBind == false` response. Can we at least log in case of an error? Or do some authorizers not respect the `allowed, message, err` format?

liggitt replied:
added logging in error case

cjcullen commented on pkg/registry/rbac/rolebinding/policybased/storage.go:
```diff
@@ -63,6 +66,12 @@ func (s *Storage) Update(ctx genericapirequest.Context, name string, obj rest.Up
 		roleBinding := obj.(*rbac.RoleBinding)
 
+		// if we're explicitly authorized to bind this role, return
+		if canBind, _, _ := rbacregistry.AuthorizeBinding(ctx, roleBinding.RoleRef, roleBinding.Namespace, s.authorizer); canBind {
```
Same question here.

liggitt committed:
Allow rolebinding/clusterrolebinding with explicit bind permission check

liggitt commented:
flake on https://github.com/kubernetes/kubernetes/issues/39604
@k8s-bot unit test this

liggitt commented:
Added integration test and updated release note.

@cjcullen can you make sure this allows users with elevated permissions in the GKE authorizer to bind RBAC roles?

@deads2k would like a quick check from you as well

cjcullen commented:
Confirmed that on a rolebinding POST, the webhook sees both the `resource: "rolebindings", verb: "create"` and `resource: "roles", verb: "bind"` SubjectAccessReviews.

cjcullen commented:
LGTM on my side. I'll let @deads2k give it a once over and apply the label.

deads2k commented on pkg/registry/rbac/escalation_check.go:
```diff
@@ -39,3 +44,46 @@ func EscalationAllowed(ctx genericapirequest.Context) bool {
+
+	switch roleRef.Kind {
+	case "ClusterRole":
+		attrs.APIGroup = roleRef.APIGroup
```
note about us not having passed validation, but being post-defaulting and conversion, so this is safe since the value won't change again and an invalid value will be rejected by validation

deads2k commented on pkg/registry/rbac/escalation_check.go:
```diff
@@ -39,3 +44,46 @@ func EscalationAllowed(ctx genericapirequest.Context) bool {
+		User: user,
+		Verb: "bind",
+		// check against the namespace where the binding is being created.
+		// this allows delegation to bind particular clusterroles within particular namespaces.
```
Add to comment about empty being expected when binding a clusterrolebinding

deads2k commented on cluster/addons/e2e-rbac-bindings/e2e-user-binding.yaml:
```diff
@@ -17,6 +17,3 @@ subjects:
 - apiVersion: rbac/v1alpha1
   kind: User
   name: kubecfg
-- apiVersion: rbac/v1alpha1
```
remove comment about kubekins in header

deads2k commented:
Just nits.  lgtm

liggitt commented:
comments addressed
